lua/orgmode/ui/virtual_indent.lua

```diff
@@ -72,22 +92,60 @@ function VirtualIndent:set_indent(bufnr, start_line, end_line, ignore_ts)
   end
 end
 
----@param bufnr? number buffer id
-function VirtualIndent:attach(bufnr)
-  bufnr = bufnr or 0
-  self:set_indent(0, 0, vim.api.nvim_buf_line_count(bufnr) - 1, true)
+--- Begins a timer to check `vim.b.org_indent_mode` if `vim.b.org_indent_mode` is not already being
+--- monitored
+function VirtualIndent:start_watch_org_indent()
+  if not self._watcher_running then
+    self._watcher_running = true
+    self._timer:start(
```

Comment:
Instead of timers, let's rely on Vim's `dictwatcheradd` and `dictwatcherdel` functions.
Those can be used only from vimscript, but there is a workaround.

Create a `dict_watcher.lua` file in the `utils` folder:

```lua
local watchers = {}
local M = {}

---@param change_dict { old?: any, new: any }
---@param key string
---@param dict table
function M.dict_changed(change_dict, key, dict)
  if watchers[key] then
    watchers[key](change_dict, key, dict)
  end
end

---@param key string
---@param callback fun(change_dict: { old?: any, new: any }, key: string, dict: table)
function M.watch_buffer_variable(key, callback)
  vim.cmd(([[
    call dictwatcheradd(b:, '%s', 'OrgmodeWatchDictChanges')
  ]]):format(key))
  watchers[key] = callback
end

---@param key string
function M.unwatch_buffer_variable(key)
  vim.cmd(([[
    call dictwatcherdel(b:, '%s', 'OrgmodeWatchDictChanges')
  ]]):format(key))
  watchers[key] = nil
end

return M
```

Add this to plugin/org.vim
```vim
function OrgmodeWatchDictChanges(dict, key, change_dict) abort
  return luaeval('require("orgmode.utils.dict_watcher").dict_changed(_A[1], _A[2], _A[3])', [a:change_dict, a:key, a:dict])
endfunction
```

Watch/unwatch for changes like this:

```lua
require('orgmode.utils.dict_watcher').watch_buffer_variable('org_indent_mode', function(change_dict)
  print('old', change_dict.old)
  print('new', change_dict.new)
end)
```

Unwatch

```lua
require('orgmode.utils.dict_watcher').unwatch_buffer_variable('org_indent_mode')
```

Reply:
So the only issue I see with dictwatcher is that there's genuine desire to deprecate it upstream:

https://github.com/neovim/neovim/issues/21469

I'll do it if you're ok with the above in mind. Just figured you should know.

Reply:
Add the link to the above issue in the `dict_watcher` file as comment to have it as a reference, and lets proceed with it. I don't think it will be deprecated that easily since it's used by few popular GUIs.

Reply:
Yep, sounds good.

Reply:
Done.